packages/image_picker/image_picker_for_web/CHANGELOG.md

```diff
@@ -1,3 +1,11 @@
+## 2.3.0
+
+* Listens to `cancel` event on file selection. When the selection is canceled:
+  * `Future<XFile?>` methods return `null`
+  * `Future<List<XFile>>` methods return an empty list.
+* Removes all code and tests mentioning `PickedFile`.
```

Comment:
This has been only around for a few weeks, but I don't want to postpone the cleanup much longer. Should I make this PR a major version bump, so it's guaranteed that nobody ends up in a state of having an out-of-date core plugin, but a "too new" web impl?)

Reply:
Sorry, I thought I'd responded to this already. Removing deprecated code is a breaking change, so requires a major version bump, otherwise you will actively break people with older versions of the app-facing package.

> but I wouldn't want to postpone this cleanup too much.

We haven't done any of the cleanup in the lower layers; we should do it bottom-up for the non-user-facing parts.

> Or can I assume that people migrated to the new API years ago, with the deprecation?

You can't assume anything that the resolver will allow to happen won't happen :)

Reply:
I checked, and we haven't even formally deprecated the old methods in the platform interface package, which makes removing the implementations here worrying since we aren't doing anything to ensure we don't call it anywhere.

I think we should deprecate the methods first in a prequel PR, and then update the min version of the platform interface this package requires to the one that deprecates them. I'll make that PR now, then we can update this one after that lands.

Reply:
@stuartmorgan the methods may not be deprecated on the `platform_interface`, but everything that returned a `PickedFile` was marked as deprecated ~2 years ago in the core plugin https://github.com/flutter/plugins/pull/4073, and eventually removed here https://github.com/flutter/packages/pull/4285, right?

AFAIK, nothing that returns a `PickedFile` is being called anymore by the plugin as of version `^1.0` (could it? yes)

(This deprecation started on the front-end, of the package, I guess :P)

As for landing this in sequence with the deprecation in the platform_interface: no problem either! :)

Reply:
> @stuartmorgan the methods may not be deprecated on the `platform_interface`, but everything that returned a `PickedFile` was marked as deprecated ~2 years ago in the core plugin [flutter/plugins#4073](https://github.com/flutter/plugins/pull/4073), and eventually removed here #4285, right?
> 
> AFAIK, nothing that returns a `PickedFile` is being called anymore by the plugin as of version `^1.0` (could it? yes)

Correct, it's not that we are calling this in the app-facing package's implementation, it's that there's nothing structural that would prevent us from calling it in the future (and as I found when I tried to do the deprecation, nothing preventing other locations in our codebase from calling it).

> (This deprecation started on the front-end, of the package, I guess :P)

There are two distinct API layers: the client-facing API, and the platform interface API. Deprecating things in the client-facing API is unrelated in general to deprecating things in the platform interface (although they happen to be more closely linked than usual here).

My concern is that the platform implementations should be implementing the platform interface API as it is expressed formally, rather than tailoring the implementations to how current clients are calling it.